src/cryptonote_core/loki_name_system.cpp

```diff
@@ -1104,50 +1281,119 @@ static bool get_txid_lns_entry(cryptonote::Blockchain const &blockchain, crypto:
+}
+
+struct replay_lns_tx
+{
+  uint64_t                              height;
+  crypto::hash                          tx_hash;
+  cryptonote::tx_extra_loki_name_system entry;
+};
+
+static std::vector<replay_lns_tx> find_lns_txs_to_replay(cryptonote::Blockchain const &blockchain, lns::mapping_record const &mapping, uint64_t blockchain_height)
+{
+  /*
+     -----------------------------------------------------------------------------------------------
+     Detach Logic: Simple Case
+     -----------------------------------------------------------------------------------------------
+     LNS Buy    @ Height 100: LNS Record={field1=a1, field2=b1, field3=c1}
+     LNS Update @ Height 200: LNS Record={field1=a2                      }
+     LNS Update @ Height 300: LNS Record={           field2=b2           }
+     LNS Update @ Height 400: LNS Record={                      field3=c2}
+
+     Blockchain detaches to height 301, the target LNS record now looks like
+                                         {field1=a2, field2=b2, field3=c1}
+
+     Our current LNS record looks like
+                                         {field1=a2, field2=b2, field3=c3}
+
+     To rebuild our record, find the closest LNS Update that is earlier than the
+     detach height. If we run out of transactions to run back to, then the LNS
+     entry is just deleted.
+
+     -----------------------------------------------------------------------------------------------
+     Detach Logic: Advance Case
+     -----------------------------------------------------------------------------------------------
+     LNS Buy    @ Height 100: LNS Record={field1=a1, field2=b1, field3=c1}
+     LNS Update @ Height 200: LNS Record={field1=a2                      }
+     LNS Update @ Height 300: LNS Record={           field2=b2           }
+     LNS Update @ Height 400: LNS Record={                      field3=c2}
+     LNS Update @ Height 500: LNS Record={field1=a3, field2=b3, field3=c3}
+     LNS Update @ Height 600: LNS Record={                      field3=c4}
+
+     Blockchain detaches to height 401, the target LNS record now looks like
+                                         {field1=a2, field2=b2, field3=c2}
+
+     Our current LNS record looks like
+                                         {field1=a3, field2=b3, field3=c4}
+
+     To get all the fields back, we can't just replay the latest LNS update
+     transactions in reverse chronological order back to the detach height,
+     otherwise we miss the update to field1=a2 and field2=b2.
+
+     To rebuild our LNS record, we need to iterate back until we find all the
+     TX's that updated the LNS field(s) until all fields have been reverted to
+     a state representative of pre-detach height.
+
+     i.e. Go back to the closest LNS record to the detach height, at height 300.
+     Next, iterate back until all LNS fields have been updated at a point in
+     time before the detach height (i.e. height 200 with field=a2).
```

Comment:
This logic doesn't seem correct to me.  If you iterate backwards like this then I think you will end up with the wrong values in this case:

```
     LNS Buy    @ Height 100: LNS Record={field1=a1, field2=b1, field3=c1}
     LNS Update @ Height 200: LNS Record={field1=a2, field2=b5           }
     LNS Update @ Height 300: LNS Record={           field2=b2           }
     LNS Update @ Height 400: LNS Record={                      field3=c2}
     LNS Update @ Height 500: LNS Record={field1=a3, field2=b3, field3=c3}
     LNS Update @ Height 600: LNS Record={                      field3=c4}
```
the algorithm will stop after applying the 200 update (because then everything has been updated), but you will have updated `field2` to `b5`, while it should be `b2`.

However I *don't* think that's what you're doing in the code: you walk back to 200 to build up the list of LNS txes, then replay forward to 400.  I think that's correct, but I suggest the example above is a better one (because it fails via backwards iteration, unlike the example you have here now).

Now all of that said: I don't see that there is a necessity to go all the way back and then walk forward again; I think you can simplify this algorithm to walk back and do a single update rather than needing to walk back and then forward with multiple updates.

Instead you just walk back and store updates to fields the *first* time you see them while walking backwards, and keep walking until you have seen all the fields that got updated in any removed blocks.  So in this example:
- 600 and 500 are removed, so my list of "need update" fields is "field1, field2, field3".
- At 400 store `update[field3] = c2` and remove field3 from my update list.
- At 300 store `update[field2] = b2` and remove field2 from my update list.
- At 200 store `update[field1] = a2` but do *not* store `update[field2]` because I already have seen an update for field2.  Now my "need updates" list is empty, so I stop walking backwards, and can do a single update from what I stored in `update` without needing to replay any txes.

Reply:
I didn't dig into the test cases, but this one I described above might be useful (to make sure we aren't doing a backwards walk/forward iteration wrong).

Reply:
Yes. I thought that it'd be nice and go replay all the transactions, the same way you would when you pop blocks and replay transactions in the blockchain.

So currently I replay all the transactions. But that is extraneous and can be done better by just recording as you said, the earliest TX that touched that field. I will work on this, if it doesn't make the testnet fork, I'll add in another PR.